### ISSUE TITLE
Action figure rig approach

### ISSUE THREAD

**Issue by eugeneloza** [eugeneloza/decoherence#391]
[Extremely complex, experimental, and might be very wrong in nature]
**Try action figure rig approach**
[ ] Skeleton, based on "parenting" of TTransformNode (TSceneTransform?)
[ ] Static bodyparts.
[ ] Load animations directly from some mocap format, supported by blender-export (for manual animations)
[ ] Maybe, just attach TSceneTransform to mocap bones
That should be dirty but simple. Moreover, this will create a specific model style, while ugly it'd be authentic  :D, support for changing clothes, unify animations for different models/actors just by scaling, overall lowering modeling skills requirements, texturing quality requirements (if the models are ugly, there's no point in high-quality texture), and support for hundreds of animations.
In other words, reduce quality in favor of quantity. Which will be beneficial both for development speed and user experience.

**Comment by eugeneloza** [comment]
Also support for simply animated models (castle-anim-frames) to be rendered with the same API (and unified reference to animations). That'd cover all possible holes in the logic.

**Comment by eugeneloza** [comment]
Just to justify myself. I'm trying to do the same thing Spine does, but in 3D. However, unused in 3D due to no mesh deformation (i.e. ugly). Can still provide for a diverse output that will cover all of my animation needs. On practical side this will just be a rig-based animations with ugly weightpaints.
![77f21b07839375b2a3f46315c8892bcb--body-reference-drawing-reference](https://user-images.githubusercontent.com/10726162/35664840-52428cbc-072c-11e8-81c3-2dc996e8124d.jpg)

**Comment by eugeneloza** [comment]
Would also provide for procedural animations, such as rag-doll by Kraft.

**Comment by eugeneloza** [comment]
I surely need to stick to some available armature format and read carefully about implementing a full-scale armature rig. Even if it won't work this way, it'll be useful.

**Comment by eugeneloza** [comment]
While tempting, ActionFigure and Armature operate in a different way. To use Armature properly I have to operate the whole mesh. To use ActionFigure I'm more than enough with just tagged `TTransformNode`, which also makes it easier and more clear how to manage (change-on-the-fly) those.

**Comment by eugeneloza** [comment]
So, we can easily "load" a normal rig (even without blender in the middle - directly read mocap format, also supporting for blender-type animations for custom models), just to attach it to `TTransformNode`s by tags. Sounds easy enough.

**Comment by ResonantExplosion** [comment]
There is a serious problem with facial expressions in this approach. The "face emotions" will require a lot of face elements, all smoothly Blended.
Practically we do not need them as badly and might be fine with only mouth open-close. Second, we don't need them to be perfectly blended, we may trigger them as discreet. Third, we may take use of facial features to create unique portraits which will also participate in emotional display.
Let's think of such head structure:
SkukkShape, ears, eyes, eye-lids, brow, mouth (lower jaw), vibros, ponytail. Those should be enough for basic diversity and emotions, _except_ for smiling. So, taking this path we forget about smiling and other lip-emoji and mimic it with vibros.
Needless to say this will become completely incompatible with predefined skeletons, forcing writing a Constructor utility to import and mix those. Also procedural animations are inavoidable.

We also need physically based body parts to handle hair, robes, tail, vibros, etc. While giving force to each bodypart we can determine the degree to which it will be managed by animation and physics.

That stuff is becoming overly complicated, I don't like it. Thou I see no other choice. Take a three month run and do it.

**Comment by ResonantExplosion** [comment]
Maybe use `TSceneTransform` to support preanimatred body parts

**Comment by eugeneloza** [comment]
@michaliskambi if you could check this issue and give your comment I'd really appreciate it (there is no need to read all the considerations above - those are just thinking aloud).

The idea is the following :

There are static or animated `TSceneTransform` instances, which are kept together by some "force" (Armature-based, there're CC0 animated armatures freely available I can write an import) to provide for character animations. They are separate, no weight-painting/skinning/deformations (I won't be able to make those properly anyway), can be procedurally animated, including by Kraft physics (e.g. ragdoll, physical hair/tail) etc.
So, that practically solves most of my "needs" in animations. Models will look ugly, but better than nothing :D Moreover, it'll be easy to customize/change clothes of the actors, one animated skeleton can be used over hundreds of different actors.
The only difference between normal rig approach and my idea - is that not mesh vertexes are attached to bones, but large `TSceneTransform`s (or maybe `TTransformNode`s - I'm not yet sure about that).

What I need to know is:

I see than CGE already has some sort of 2D/3D armature support, and of course it is much better to use/enchance this implementation instead of inventing the wheel :) Could you direct me to the right classes/documentation to check how it's done and see if I can use a ready solution? Anyway I bet you have some plans to, can you point me in the right direction, so that this work could be useful not only for my project, but also for CGE?

**Comment by eugeneloza** [comment]
Certainly that's not urgent :)
I'm still deep in the "planning" stage.
However, I'm already sure that I'll use this approach in some future.

**Comment by michaliskambi** [comment]
(Sorry for answering with delay -- I missed the fact that I was mentioned. I improved my mail filters now so it will not happen again. :) )

In short: for skeletons where each part is rigid, I advice to just build a hierarchy of X3D `Transform` nodes (`TTransformNode` in Pascal). It's simple, optimized, and proven to work (we animate quite complicated Spine skeletons this way in "Escape from the Universe" and "The Unholy Society"). You can attach anything as children of `TTransformNode`, so it is quite flexible.

Longer version:

I stared writing an answer, but it got awfully long. So instead I pasted it into the documentation: https://castle-engine.io/x3d_implementation_interpolation.php has now a huge section "_How to animate things using X3D nodes_". Enjoy, I hope that this is useful, and please ask if you have any questions :)

Some things I do not discuss there:

- Although you *can* build a hierarchy of `TCastleTransform` instead of a hierarchy of `TTranformNode`, but it's probably more comfortable to keep the whole animation inside a single scene (so, make a hierarchy of `TTranformNode`). This way the animation can be controlled as one, by a simple call to e.g. `TCastleSceneCore.PlayAnimation` and friends.

- As for physics: Note that right now only whole `TCastleTransform` can be a rigid body for physics. But you cannot make ragdoll anyway now, because we do not support "constraints". I plan to extend our Kraft integration to cover this, and to have rigid bodies available also for particular Transform nodes inside scene. In any case, if you think about ragdolls -- they will require extending CGE anyway, whether you build a hierarchy of `TCastleTransform` or `TTranformNode`. It *will* happen of course :)

**Comment by eugeneloza** [comment]
Thanks a lot, Michalis! It's no problem with the delay :) I'm still deep in "planning phase", however, I'm already itching to try something as the idea seems to be quiet set already.
I've thought of implementing the constrains myself (including breakable constrains like shattering a skeleton to pieces :)). The idea behind using `TCastleTransform` was that I could have a blender-animated head and hands and armature-animated body. However, at the moment I see that there is hardly any sense in merging the animations and bodyparts animations should be either prerecorded by armature or being procedural. The idea behind all that is that the character may play "idle" animation, but rotate his/her head towards the speaker, open the mouth for speaking, etc., etc. Ragdoll is a "bonus" and not actually something needed all that badly :).
I've printed your reply and the article to read it carefully tomorrow, hopefully it'll make things much clearer to me (or I'll have some more specific questions then).

**Comment by michaliskambi** [comment]
The Seamless3d ( http://www.seamless3d.com/ ) software is also nicely connected to X3D and H-Anim (I was actually testing our H-Anim implementation on models produced by it). And they list among their features _"Imports & Exports BVH motion capture files"_ . They have some links on http://www.seamless3d.com/links/index.html#bvh too.

**Comment by michaliskambi** [comment]
P.S. I improved the text on https://castle-engine.io/x3d_implementation_interpolation.php a bit, and added there example showing how to build it in Pascal, if necessary.

**Comment by eugeneloza** [comment]
I haven't studied all the links provided yet, thou it already seems a bit different than what I need. At this point I want the "model" to run several animations separately. At least mocap+procedural animations (e.g. some body or clothes part like tail or cloak may follow physics, eyes may blink independently of movement or idle animation).
I was thinking of "not using" `TTimeSensorNode`, but managing the `TTransformNodes` transform manually (i.e. copying it from the rig in case of a predefined animation for the bodypart or using some procedural animation (like blinking) in case there is no rig for the bodypart or if a different animation is requested (e.g. while playing "walk" animation with head facing forward, we may force the `Actor` to turn his head towards camera, or forcing the actor to close the eyes when taking damage to override default blinking animation)).
But this is very theoretical at the moment. I'm still studying the specifications you've provided.

**Comment by eugeneloza** [comment]
On re-reading the article again and again several times, I seem to have finally understood the approach. Yes, that looks like a way to go (using multiple `TTimeSensorNode`s some of which would be procedurally generated).

**Comment by ResonantExplosion** [comment]
So, the plan is clear.
1. Parse asf/ams files
2. Create the animations nodes + think about best storage container variant.
3. Make a test model
4. Attach model TTransformNodes by tag to animation modes.
5. Test single animations
6. Test animation blending
And we're done with the critical part of the issue.
Procedural animations and facial expressions are a bonus and can be postponed for indefinite time. We need to keep those in mind, but first things first. Otherwise this issue will never leave planning stage.

**Comment by ResonantExplosion** [comment]
And don't forget about bvh files, as blender can export those

**Comment by michaliskambi** [comment]
Note that some of the links I sent suggested that there is already software that converts BVH to X3D:) 


And you can possibly convert ASF to BVH through Blender, as Blender can probably import ASF ( https://wiki.blender.org/index.php/Extensions:2.6/Py/Scripts/Import-Export/Acclaim_Importer ). Assuming that Seamless 3d cannot convert straight from ASF to X3D.

**Comment by eugeneloza** [comment]
Yes, but loading them directly is a nice feature to have in the Engine too :) Moreover, converting software does so inefficiently. E.g. converting ASF to BVH makes it 1Gb -> 4Gb as it forces discrete interpolation (which is another drawback).
/still reading through the format documentation:)/

**Comment by eugeneloza** [comment]
/Finished with reading through most of the documentation. Looks really cool. Now to "digest" it and may start putting into practice. Yes, keeping as close as possible to standard solutions is a way to go./
It doesn't seem there is Acclaim exporter available for blender, BVH only, so it's essential to have at least those two formats supported and converted/saved to X3D H-Anim standard.

**Comment by ResonantExplosion** [comment]
Thinking aloud: we need at least:
Humanoid skeleton (mocap available)
Worm skeleton (ik easy)
Helicopter skeleton (ik easy)
Blob skeleton (ik easy)
Hydra skeleton (ik easy)
Slug skeleton (ik easy)
Wisp skeleton (easy)
Wortex skeleton (easy)
Rodent skeleton (hard, but may be skipped)
Bird skeleton (hard, but may be skipped)
Quadrupled skeleton (the trickiest thing, inavoidable)
Scorpion/spider/ant skeleton (should be not too difficult)
Biped skeleton (might get tricky)
Giant skeleton (???)
Jabb skeleton (???)
Jumper skeleton (kinda kangoroo, very tricky)

**Comment by ResonantExplosion** [comment]
Sphere skeleton (easy)

**Comment by ResonantExplosion** [comment]
Centipede skeleton (ik easy)
